core/services/postgres/lease_lock.go

```diff
@@ -0,0 +1,211 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"github.com/smartcontractkit/sqlx"
+)
+
+// LeaseLock handles taking an exclusive lease on database access. This is not
+// enforced by any database primitives, but rather voluntarily respected by
+// other instances of the Chainlink application.
+//
+// Chainlink is designed to run as a single instance. Running multiple
+// instances of Chainlink on a single database at the same time is not
+// supported and likely to lead to strange errors and possibly even data
+// integrity failures.
+//
+// With that being said, a common use case is to run multiple Chainlink
+// instances in failover mode. The first instance will take some kind of lock
+// on the database and subsequent instances will wait trying to take this lock
+// in case the first instance disappears or dies.
+//
+// Traditionally Chainlink has used an advisory lock to manage this. However,
+// advisory locks come with several problems, notably: - Postgres does not
+// really like it when you hold locks open for a very long time (hours/days).
+// It hampers certain internal cleanup tasks and is explicitly discouraged by
+// the postgres maintainers - The advisory lock can silently disappear on
+// postgres upgrade - Advisory locks do not play nicely with pooling tools such
+// as pgbouncer - If the application crashes, the advisory lock can be left
+// hanging around for a while (sometimes hours) and can require manual
+// intervention to remove it
+//
+// For this reason, we now use a database leaseLock instead, which works as
+// such: - Have one row in a database which is updated periodically with the
+// client ID - CL node A will run a background process on start that updates
+// this e.g. once per second - CL node B will spinlock, checking periodically
+// to see if the update got too old. If it goes more than, say, 5s without
+// updating, it assumes that node A is dead and takes over. Now CL node B is
+// the owner of the row and it updates this every second - If CL node A comes
+// back somehow, it will go to take out a lease and realise that the database
+// has been leased to another process, so it will panic and quit immediately
+type LeaseLock interface {
+	TakeAndHold() error
+	Release()
+	ClientID() uuid.UUID
+}
+
+var _ LeaseLock = &leaseLock{}
+
+type leaseLock struct {
+	// TODO: Use a "master" application parent ctx that is cancelled on stop?
```

Comment:
This is optional but if we hit ctrl-c while the app is still booting it can take a while to exit because it doesnt cancel contexts for queries that are issued during Start calls